### PR TITLE
fix(mercury): timing condition where disconnect did not close websockets

### PR DESCRIPTION
**Pull request by gannons**
# Pull Request Template

## Description

We've a E2E test app to validate the eDiscovery Download Manager. But we found the app was not shutting down without a `process.exit(..)` call. Using [wtfnode](https://www.npmjs.com/package/wtfnode) I identified the mercury socket as being the cause. It was not closed and this meant the app remained running. I "fixed" this with

    await spark.internal.mercury.disconnect()

in the test's `after` section. But, intermittently, the app would still hang and at the same time I'd see a `unhandledRejection` with an `undefined` error.

I tracked the rejection down to [mercury's abort handling](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js#L318) which gets called by the [abort event in the backoff module](https://github.com/MathieuTurcotte/node-backoff/blob/master/lib/function_call.js#L120) in turn triggered by [spark.internal.mercury.disconnect()](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js#L77).

Debugging into the code I noticed that `disconnect` could be called while a `socket.open(..)` was in progress. That meant the socket would never be closed as, at that time, the [mercury socket](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js#L80) was undefined.

To fix this I'm updating the code to set `this.socket` just before calling `socket.open(..)`. That means the disconnect can close the socket. i.e. Preventing my app from hanging.

**Caveat:** I'm no mercury expert and would appreciate the core team taking a careful look at this PR.

**Additional:** I see another bug in the code related to `spark.internal.mercury.disconnect()`. If the [socket times out when closed](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/%40webex/internal-plugin-mercury/src/socket/socket-base.js#L128) it causes the follow close event

          resolve(this.onclose({
            code: 1000,
            reason: 'Done (forced)'
          }));

The [mercury code](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js#L24) checks for the reason `Done (forced)` and if it detects it then the [socket will be re-connected](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js#L409). It seems to consider this case a _normalReconnectReasons_ but a [normal close](https://github.com/webex/webex-js-sdk/blob/master/packages/node_modules/%40webex/internal-plugin-mercury/src/socket/socket-base.js#L139), which translates to

        this.onclose({
            code: 1000,
            reason: 'Done'
          }));

does not cause a re-connect.

This means when you call

    spark.internal.mercury.disconnect()

and it closes the socket, but there is a time out, the socket is re-opened. That is not intuitive and _seems_ like a mistake. But it also _looks very_ deliberate in the code. I can't figure out why the SDK would do this which means

 a) Its a long standing bug or
 b) I'm misunderstanding something

To fix this I'd like to removed `Done (forced)` from `normalReconnectReasons`. But I'm wary of breaking something I'm missing. Can one of the core team have a look?

Fixes # (issue)

## Type of Change

Please delete options that are not relevant.

- [x] Bug fix (non-breaking change which fixes an issue)

## Test Coverage

Unit test updated and unit test added.

**Test Configuration**:
* SDK Version
* Node/Browser Version
* NPM Version

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js [hunk 1]
```diff
@@ -167,6 +167,12 @@ const Mercury = WebexPlugin.extend({
 
     Promise.all([this._prepareUrl(socketUrl), this.webex.credentials.getUserToken()])
       .then(([webSocketUrl, token]) => {
+        if (!this.backoffCall) {
+          this.logger.info('mercury: prevent socket open when backOffCall no longer defined');
+
+          return Promise.reject(new Error('mercury: prevent socket open when backOffCall no longer defined due to abort?'));
```
Comment (gannons) [inline packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js]: Handling the scenario where `_attemptConnection` was called, the `Promise.all(..)` started, a `disconnect()` runs, then the `Promise.all(..)` resolved. Without this check the code would still open the socket after a disconnect.

File: packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js [hunk 2]
```diff
@@ -315,7 +324,7 @@ const Mercury = WebexPlugin.extend({
 
       call.on('abort', () => {
         this.logger.info('mercury: connection aborted');
-        reject();
+        reject(new Error('Mercury Connection Aborted'));
```
Comment (gannons) [inline packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js]: Include an error to help identify the problem. Much better than undefined.

File: packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js [hunk 3]
```diff
@@ -3,6 +3,12 @@
  */
 
 import {EventEmitter} from 'events';
+
+import {checkRequired} from '@webex/common';
+import {safeSetTimeout} from '@webex/common-timers';
+import {defaults, has, isObject} from 'lodash';
+import uuid from 'uuid';
```
Comment (gannons) [inline packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js]: eslint check did this for some reason

File: packages/node_modules/@webex/internal-plugin-mercury/test/unit/spec/mercury.js [hunk 4]
```diff
@@ -504,20 +504,54 @@ describe('plugin-mercury', () => {
...
+              // By this time backoffCall and mercury socket should be defined by the
+              // 'connect' call
+              assert.isDefined(mercury.backoffCall, 'Mercury backoffCall is not defined');
+              assert.isDefined(mercury.socket, 'Mercury socket is not defined');
+              // Calling disconnect will abort the backoffCall, close the socket, and
+              // reject the connect
               mercury.disconnect();
+              assert.isUndefined(mercury.backoffCall, 'Mercury backoffCall is still defined');
+              // The socket will never be unset (which seems bad)
+              assert.isDefined(mercury.socket, 'Mercury socket is not defined');
```
Comment (gannons) [inline packages/node_modules/@webex/internal-plugin-mercury/test/unit/spec/mercury.js]: Assumed the `socket` would have been blanked but that is not the case. Slightly worrying for re-connect / multiple connect scenarios.

**Comment by gannons** [comment]
FYI - when running the mercury tests locally I found them to be a flakey. But they seem to be passing fairly well _today_. i.e. I had no failures so far.

The issue I tend to hit though is `Error: Timeout of 30000ms exceeded`.

File: packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js (hunk 3 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js]: This usually happens based on the import type, but it is strange that it separated `events` from the rest. We can disregard this formatting change as it's not impacting.

File: packages/node_modules/@webex/internal-plugin-mercury/test/unit/spec/mercury.js (hunk 4 above)
Comment (InteractiveTimmy) [reply packages/node_modules/@webex/internal-plugin-mercury/test/unit/spec/mercury.js]: In the case of this test, the socket remains active after the connection is no longer valid? Just wanting to confirm based on the implemented flow.

Comment (gannons) [reply packages/node_modules/@webex/internal-plugin-mercury/test/unit/spec/mercury.js]: The socket will be closed and inactive. If another call is made to `connect()` it'll set the `socket` variable to a new instance of `Socket`. Initially I assumed the variable would be set to `undefined` to match the start state. i.e. In case the code used the existence of `socket` for any logic.

**Comment by gannons** [comment]
@InteractiveTimmy any opinions on additional bug I mentioned above? i.e. Timeout on socket close causes a re-connect.

**Comment by InteractiveTimmy** [comment]
> @InteractiveTimmy any opinions on additional bug I mentioned above? i.e. Timeout on socket close causes a re-connect.

In reference to the timeout error, are you running `npm test -- --packages @webex/internal-plugin-mercury`? We will create a bug case on our end to investigate that error further. To help scope the bug case, do you know if the same consistency occurs on our master branch?

**Comment by gannons** [comment]
> In reference to the timeout error, are you running npm test -- --packages @webex/internal-plugin-mercury?

I have noticed an intermittent timeout error with the tests. But, here, I'm referring to the section in the PR description. i.e.

> **Additional**: I see another bug in the code related to spark.internal.mercury.disconnect(). If the socket times out when closed it causes the follow close event  ........

**tldr:** If the socket `close` times out after 2 seconds it fires an event with the reason `Done (forced)`. When mercury detects that event it _re-connects_. So a disconnect can end up causing a re-connect. i.e. Rather than disconnect mercury, if there is a timeout, it re-connects it.

File: packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js [hunk 5]
```diff
@@ -184,10 +190,13 @@ const Mercury = WebexPlugin.extend({
           options = {...options, ...this.webex.config.defaultMercuryOptions};
         }
 
+        // Set the socket before opening it. This allows a disconnect() to close
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js]: Nice catch!

File: packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js (hunk 1 above)
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js]: Is there a reason why the error message is slightly different than the one just above it? Could we just reuse the string?

File: packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js [hunk 6]
```diff
@@ -167,6 +167,12 @@ const Mercury = WebexPlugin.extend({
 
     Promise.all([this._prepareUrl(socketUrl), this.webex.credentials.getUserToken()])
       .then(([webSocketUrl, token]) => {
+        if (!this.backoffCall) {
+          this.logger.info('mercury: prevent socket open when backOffCall no longer defined');
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js]:
```suggestion
          this.logger.info('mercury: prevent socket open when backoffCall no longer defined');
```
Capitalization didn't match the actual property.

File: packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js [hunk 7]
```diff
@@ -98,9 +100,17 @@ export default class Socket extends EventEmitter {
    */
   close(options) {
     return new Promise((resolve, reject) => {
-      this.logger.info('socket: closing');
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js]: Could we leave this log in?

File: packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js [hunk 8]
```diff
@@ -98,9 +100,17 @@ export default class Socket extends EventEmitter {
...
       const socket = sockets.get(this);
 
+      if (!socket) {
+        // Open has not been called yet so there is no socket to close
```
Comment (adamweeks) [inline packages/node_modules/@webex/internal-plugin-mercury/src/socket/socket-base.js]: Let's add an info log with the message of this comment here

File: packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js (hunk 1 above)
Comment (gannons) [reply packages/node_modules/@webex/internal-plugin-mercury/src/mercury.js]: No good reason - I've updated to make consistent

**Comment by gannons** [comment]
Any update on this PR? I pushed a new commit to address the logging issue pointed out by @adamweeks.

The new failure in _samples_ looks to be a timing condition rather than related to the latest change.

> Timed-out waiting for screenshare to start

Imagine a re-run would resolve.
